### PR TITLE
Central storage of session model data

### PR DESCRIPTION
**Pull request by HughePaul**

### CODE REVIEW

File: test/spec.wizard-model.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    it('should be an instance of SessionModel', () => {
+        wizardModel.should.be.an.instanceOf(SessionModel);
+    });
+
+    describe('constructor', () => {
+        it('should throw an error if no options are provided', () => {
+            expect(() => new WizardModel()).to.throw();
+        });
+
+        it('should throw an error if no journeyModel is provided in options', () => {
```
Comment (merlinc) [inline test/spec.wizard-model.js]: Could we also get a test that checks to see that journeyModel is an instance of SessionModel?

File: lib/wizard-model.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            throw new Error('wizard-model - fields must be defined');
+        }
+
+        super(attrs, options);
+
+        journeyModel.registerModel(this);
+    }
+
+    getJourneyKeys() {
+        return _.pick(
```
Comment (merlinc) [inline lib/wizard-model.js]: Could we get some empty object tests for this?

File: lib/wizard-model.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            _.mapObject(this.options.fields, f => f.journeyKey),
+            _.isString
+        );
+    }
+
+    getDefaults() {
+        return _.pick(
```
Comment (merlinc) [inline lib/wizard-model.js]: Could we get some empty object tests for this?

File: test/spec.journey-model.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
+'use strict';
+
+const JourneyModel = require('../lib/journey-model');
+const SessionModel = require('../lib/model');
+
+describe('session model', () => {
```
Comment (merlinc) [inline test/spec.journey-model.js]: * journey-model

File: test/spec.journey-model.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        it('should not add duplicate model keys to the model list', () => {
+            let session = {
+                test: {
+                    'registered-models': [ 'test1' ]
+                }
+            };
+            let modelStub = {
+                options: { key: 'test1' }
+            };
+            let journeyModel = new JourneyModel(null, { key: 'test', session });
```
Comment (merlinc) [inline test/spec.journey-model.js]: Is there any way to make this test clearer? It took a little bit of thinking to separate out the `test1` key lines from the `test` key line.

File: lib/controller/mixins/session-model.js [hunk 6]
```diff
@@ -10,17 +10,19 @@ module.exports = Controller => class extends Controller {
...
         if (req.sessionModel) {
             req.sessionModel.destroy();
         }
 
-        req.sessionModel = new Model(null, {
+        req.sessionModel = new WizardModel(null, {
             session: req.session,
-            key: 'hmpo-wizard-' + req.form.options.name
+            key: 'hmpo-wizard-' + req.form.options.name,
+            journeyModel: req.journeyModel,
```
Comment (HughePaul) [inline lib/controller/mixins/session-model.js]: WizardModel will throw an error, express will catch and pass to the errorHandler

File: lib/wizard-model.js (hunk 2 above)
Comment (HughePaul) [inline lib/wizard-model.js]: added

File: lib/wizard-model.js (hunk 3 above)
Comment (HughePaul) [inline lib/wizard-model.js]: added

File: lib/wizard-model.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            v => v !== undefined
+        );
+    }
+
+    getSessionData() {
+        let defaults = this.getDefaults();
+        let wizardData = super.getSessionData();
+        let journeyData = _.mapObject(
+            this.getJourneyKeys(),
+            f => this.options.journeyModel.get(f)
```
Comment (HughePaul) [inline lib/wizard-model.js]: I have changed this, and others, to be more descriptive

File: test/spec.wizard-model.js (hunk 1 above)
Comment (HughePaul) [inline test/spec.wizard-model.js]: have added a test for this and fields

**Comment by oliverweighell** [comment]
- Like the idea of registering models.
- Like that all field values remain easily gettable from sessionModel (not just the current form vals) 
- Undecided about how tied this is to the journeyModel. I know that is our biggest use case but what if we wanted to collect session data that exists outside of the 'passport' journey? An example might be the data we need for the feedback form (although potentially that is journeyModel data) or maybe a module with its own ideas about where things should be stored on the session. Is this too much flexibity, or am I missing something in the implementation?
- Would be good to get an explanation of how this can be used at the field level and also in terms of extending by registering more models.
- Fwiw, have pushed up what I'd played around with yesterday in case of any value (not working code). Includes a flag (ignoreDefault) to allow fields to store outside of the journey model if needed. https://github.com/UKHomeOffice/passports-form-wizard/compare/feature/custom-session-storage-for-fields

File: lib/journey-model.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+class JourneyModel extends SessionModel {
+
+    registerModel(model) {
+        let modelNames = this.get('registered-models') || [];
+        if (!_.contains(modelNames, model.options.key)) {
+            modelNames.push(model.options.key);
+            this.set('registered-models', modelNames);
+        }
+        this.registeredModels = this.registeredModels || [];
+        this.registeredModels.push(model);
```
Comment (oliverweighell) [inline lib/journey-model.js]: Not immediately clear what the difference is. Maybe this.currentModels?

File: lib/wizard-model.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        let journeyData = _.mapObject(
+            this.getJourneyKeys(),
+            journeyKey => this.options.journeyModel.get(journeyKey)
+        );
+        return _.defaults(journeyData, wizardData, defaults);
+    }
+
+    resetSessionData() {
+        super.resetSessionData();
+        this.attributes = this.getSessionData();
```
Comment (oliverweighell) [inline lib/wizard-model.js]: This is to reseed with defaults and journey data? Might be worth a comment to explain.

Comment (HughePaul) [inline lib/wizard-model.js]: Yes it is, as that isn't being reset. It does need a comment and/or docs

File: README.md [hunk 10]
```diff
@@ -101,6 +101,31 @@ Any of these options can also be provided as a third argument to the wizard to c
 
 Remaining field options documentation can be found in the hmpo-template-mixins [README](https://github.com/UKHomeOffice/passports-template-mixins#options-1).
 
+## Central journey storage
+To facilitate sharing form values between wizards in the same journey a field cat be specified to save into the `journeyModel` instead of the `sessionModel` using the `journeyKey` property:
```
Comment (oliverweighell) [inline README.md]: can

File: lib/journey-model.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+class JourneyModel extends SessionModel {
+
+    registerModel(model) {
+        let modelNames = this.get('registered-models') || [];
+        if (!_.contains(modelNames, model.options.key)) {
+            modelNames.push(model.options.key);
+            this.set('registered-models', modelNames);
+        }
+        this.currentModels = this.currentdModels || [];
```
Comment (oliverweighell) [inline lib/journey-model.js]: currentdModels

Comment (merlinc) [inline lib/journey-model.js]: Ooh. Good catch. Why isn't this failing any tests?

Comment (HughePaul) [inline lib/journey-model.js]: registering needs to be tested twice, it defaults to empty so it works first time.

Comment (HughePaul) [inline lib/journey-model.js]: I have added a test and fix the typo

**Comment by merlinc** [comment]
LGTM
